File: hmvvm/src/main/java/com/see99/hmvvm/dagger/module/ActivityModule.java
package com.see99.hmvvm.dagger.module;

import android.app.Application;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;


import com.see99.hmvvm.Supplier;
import com.see99.hmvvm.ViewModelProviderFactory;
import com.see99.hmvvm.base.BaseActivity;
import com.see99.hmvvm.model.http.HttpManager;
import com.see99.hmvvm.ui.login.LoginViewModel;
import com.see99.hmvvm.ui.main.MainViewModel;


import dagger.Module;
import dagger.Provides;

@Module
public class ActivityModule {

    @Provides
    LoginViewModel providesLoginViewModel(HttpManager httpManager, BaseActivity<?,?> activity, Application application){
        Supplier<LoginViewModel> supplier = () -> new LoginViewModel(httpManager);
        ViewModelProviderFactory<LoginViewModel> factory = new ViewModelProviderFactory(LoginViewModel.class, supplier);
        return new ViewModelProvider(activity, factory).get(LoginViewModel.class);
//        return new ViewModelProvider(activity,ViewModelProvider.AndroidViewModelFactory.getInstance(application)).get(LoginViewModel.class);
    }
    @Provides
    MainViewModel providesMainViewModel(HttpManager httpManager, BaseActivity<?,?> activity, Application application){
        Supplier<MainViewModel> supplier = () -> new MainViewModel(httpManager);
        ViewModelProviderFactory<LoginViewModel> factory = new ViewModelProviderFactory(MainViewModel.class, supplier);
        return new ViewModelProvider(activity, factory).get(MainViewModel.class);
//        return new ViewModelProvider(activity,ViewModelProvider.AndroidViewModelFactory.getInstance(application)).get(LoginViewModel.class);
    }

    @Provides
    BaseActivity provideActivity(BaseActivity<?,?> activity){
        return activity;
    }

    @Provides
    Toast provideToast(BaseActivity activity){
        return Toast.makeText(activity,"测试",Toast.LENGTH_LONG);
    }


}

File: hmvvmkotlin/src/main/java/com/see99/hmvvmkotlin/dagger/module/ActivityModule.kt
package com.see99.hmvvmkotlin.dagger.module

import androidx.lifecycle.ViewModelProvider
import com.see99.hmvvmkotlin.base.BaseActivity
import com.see99.hmvvmkotlin.dagger.scope.ActivityScope
import com.see99.hmvvmkotlin.ui.ViewModelProviderFactory
import com.see99.hmvvmkotlin.ui.login.LoginViewModel
import com.see99.hmvvmkotlin.ui.main.MainViewModel
import dagger.Module
import dagger.Provides
import javax.inject.Provider

@Module
class ActivityModule {


//    @ActivityScope
//    @Provides
//    fun provideActivity(activity:BaseActivity<*,*>) = activity


    @ActivityScope
    @Provides
    fun provideLoginViewModel(activity:BaseActivity<*,*>):LoginViewModel{
        var provider = Provider<LoginViewModel>{ LoginViewModel() }
        var factory = ViewModelProviderFactory<LoginViewModel>(LoginViewModel::class.java,provider)
        return ViewModelProvider(activity,factory).get(LoginViewModel::class.java)
    }

    @ActivityScope
    @Provides
    fun provideMainViewModel(activity:BaseActivity<*,*>):MainViewModel{
        var provider = Provider<MainViewModel>{ MainViewModel() }
        var factory = ViewModelProviderFactory<MainViewModel>(MainViewModel::class.java,provider)
        return ViewModelProvider(activity,factory).get(MainViewModel::class.java)
    }


}
